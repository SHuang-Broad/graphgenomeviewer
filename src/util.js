// simple handcoded GFA parser
export function parseGFA(file) {
  const graph = { nodes: [], links: [], paths: [] }
  for (const line of file.split('\n')) {
    if (line.startsWith('S')) {
      const [, name, sequence, ...rest] = line.split('\t')
      const tags = {}
      for (let i = 0; i < rest.length; i++) {
        const [name, type, val] = rest[i].split(':')
        if (type === 'i') {
          tags[name] = +val
        } else if (type === 'Z') {
          tags[name] = val
        }
      }
      graph.nodes.push({ id: name, sequence, tags })
    } else if (line.startsWith('L')) {
      const [, source, strand1, target, strand2, cigar, ...rest] = line.split('\t')
      const tags = {}
      for (let i = 0; i < rest.length; i++) {
        const [name, type, val] = rest[i].split(':')
        if (type === 'i') {
          tags[name] = +val
        } else if (type === 'Z') {
          tags[name] = val
        }
      }
      graph.links.push({ source, target, strand1, strand2, cigar, tags })
    } else if (line.startsWith('P')) {
      const [, name, path, ...rest] = line.split('\t')

      graph.paths.push({ name, path, rest })
    }
  }
  return graph
}

// adapted from https://observablehq.com/@mbostock/saving-svg
export function serialize(svg) {
  const xmlns = 'http://www.w3.org/2000/xmlns/'
  const xlinkns = 'http://www.w3.org/1999/xlink'
  const svgns = 'http://www.w3.org/2000/svg'

  svg = svg.cloneNode(true)
  const fragment = `${window.location.href}#`
  const walker = document.createTreeWalker(svg, NodeFilter.SHOW_ELEMENT, null, false)
  while (walker.nextNode()) {
    for (const attr of walker.currentNode.attributes) {
      if (attr.value.includes(fragment)) {
        attr.value = attr.value.replace(fragment, '#')
      }
    }
  }
  svg.setAttributeNS(xmlns, 'xmlns', svgns)
  svg.setAttributeNS(xmlns, 'xmlns:xlink', xlinkns)
  const serializer = new window.XMLSerializer()
  const string = serializer.serializeToString(svg)
  return new Blob([string], { type: 'image/svg+xml' })
}

// Given a GFA graph with sequence nodes ('S' tags), it breaks the S tags into
// multiple nodes depending on how long the sequence is, which gives the graph
// an organic look when the layout algorithm is applied
export function reprocessGraph(G, chunkSize) {
  const Gp = { nodes: [], links: [] }

  const seen = {}
  for (let i = 0; i < (G.paths || {}).length; i++) {
    const path = G.paths[i]
    const pathNodes = path.path.split(',')
    for (let j = 0; j < pathNodes.length - 1; j++) {
      const curr = `${pathNodes[j]}_${pathNodes[j + 1]}`
      if (!seen[curr]) {
        seen[curr] = [path.name]
      } else {
        seen[curr].push(path.name)
      }
    }
  }

  for (let i = 0; i < G.nodes.length; i++) {
    const { id, sequence, ...rest } = G.nodes[i]
    const nodes = []
    const length = sequence === '*' ? rest.tags.LN : sequence.length

    // break long sequence into multiple nodes, for organic layout
    nodes.push({ ...rest, id: `${id}-start` })
    for (let i = chunkSize; i < length - chunkSize; i += chunkSize) {
      nodes.push({ ...rest, id: `${id}-${i}` })
    }
    nodes.push({ ...rest, id: `${id}-end` })

    // recreate links
    for (let j = 0; j < nodes.length - 1; j++) {
      const source = nodes[j].id
      const target = nodes[j + 1].id
      Gp.links.push({
        ...rest,
        source,
        target,
        id,
        length,
        sequence,
        linkNum: i,
      })
    }
    Gp.nodes = Gp.nodes.concat(nodes)
  }
  for (let i = 0; i < G.links.length; i++) {
    const { strand1, strand2, source, target, ...rest } = G.links[i]
    const paths = seen[`${source}${strand1}_${target}${strand2}`] || []
    const loop = source === target

    // enumerates cases for which end of source connects to
    // which end of the target
    const link = {
      source: `${source}-${strand1 === '+' ? 'end' : 'start'}`,
      target: `${target}-${strand2 === '+' ? 'start' : 'end'}`,
      ...rest,
    }
    if (loop) {
      link.loop = true
    }
    if (paths.length) {
      link.paths = paths
    }
    Gp.links.push(link)
  }
  return Gp
}

// source https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
// returns an array that contains groupings of xs by attribute key
function groupByArray(xs, key) {
  return xs.reduce(function (rv, x) {
    const v = key instanceof Function ? key(x) : x[key]
    if (v !== undefined) {
      const el = rv.find(r => r && r.key === v)
      if (el) {
        el.values.push(x)
      } else {
        rv.push({ key: v, values: [x] })
      }
    }
    return rv
  }, [])
}

// connects successive start->end to a path
// param edges: {source:{x,y}, target:{x,y}}[]
function makePath(edges) {
  const path = []
  let i = 0
  for (; i < edges.length; i++) {
    const st = edges[i]
    path.push([st.source.x, st.source.y])
  }
  const last = edges[i - 1]
  path.push([last.target.x, last.target.y])
  return path
}

// groups the edges data structure by the linkNum attribute
export function generatePaths(edges, graph) {
  const ret = groupByArray(edges, 'linkNum')
  return ret.map(entry => {
    return {
      links: makePath(entry.values),
      original: graph[entry.key],
    }
  })
}

export function generateEdges(links, graph) {
  const result = []
  for (let i = 0; i < links.length; i++) {
    const link = links[i]
    const original = graph[i]
    if (!original.id) {
      result.push({
        links: [
          [link.source.x, link.source.y],
          [link.target.x, link.target.y],
        ],
        original,
      })
    }
  }
  return result
}
