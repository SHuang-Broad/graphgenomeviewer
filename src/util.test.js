import { parseGFA } from './util'
import * as fs from 'fs'
test('parses a GFA file', () => {
  const files = ['public/example1.gfa2']
  files.forEach(filename => {
    const contents = fs.readFileSync(filename, 'utf8')
    expect(parseGFA(contents)).toMatchSnapshot()
  })
})
