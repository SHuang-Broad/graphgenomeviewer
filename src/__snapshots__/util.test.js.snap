// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses a GFA file 1`] = `
Object {
  "header": Array [
    Object {
      "VN": "2.0",
    },
  ],
  "links": Array [
    Object {
      "cigar": "10M",
      "source": "1",
      "strand1": "+",
      "strand2": "+",
      "tags": Object {},
      "target": "2",
    },
    Object {
      "cigar": "10M",
      "source": "1",
      "strand1": "-",
      "strand2": "",
      "tags": Object {},
      "target": "19",
    },
    Object {
      "cigar": "10M",
      "source": "10",
      "strand1": "+",
      "strand2": "0",
      "tags": Object {},
      "target": "3",
    },
    Object {
      "cigar": "10M",
      "source": "10",
      "strand1": "-",
      "strand2": "0",
      "tags": Object {},
      "target": "4",
    },
    Object {
      "cigar": "10M",
      "source": "11",
      "strand1": "-",
      "strand2": "1",
      "tags": Object {},
      "target": "6",
    },
    Object {
      "cigar": "10M",
      "source": "11",
      "strand1": "+",
      "strand2": "1",
      "tags": Object {},
      "target": "9",
    },
    Object {
      "cigar": "10M",
      "source": "12",
      "strand1": "+",
      "strand2": "2",
      "tags": Object {},
      "target": "9",
    },
    Object {
      "cigar": "10M",
      "source": "12",
      "strand1": "-",
      "strand2": "-",
      "tags": Object {},
      "target": "18",
    },
    Object {
      "cigar": "10M",
      "source": "16",
      "strand1": "+",
      "strand2": "+",
      "tags": Object {},
      "target": "20",
    },
    Object {
      "cigar": "10M",
      "source": "16",
      "strand1": "-",
      "strand2": "-",
      "tags": Object {},
      "target": "22",
    },
    Object {
      "cigar": "10M",
      "source": "18",
      "strand1": "+",
      "strand2": "+",
      "tags": Object {},
      "target": "19",
    },
    Object {
      "cigar": "10M",
      "source": "18",
      "strand1": "-",
      "strand2": "-",
      "tags": Object {},
      "target": "23",
    },
    Object {
      "cigar": "10M",
      "source": "2",
      "strand1": "+",
      "strand2": "+",
      "tags": Object {},
      "target": "5",
    },
    Object {
      "cigar": "10M",
      "source": "2",
      "strand1": "+",
      "strand2": "+",
      "tags": Object {},
      "target": "5",
    },
    Object {
      "cigar": "10M",
      "source": "2",
      "strand1": "-",
      "strand2": "-",
      "tags": Object {},
      "target": "8",
    },
    Object {
      "cigar": "10M",
      "source": "20",
      "strand1": "+",
      "strand2": "+",
      "tags": Object {},
      "target": "21",
    },
    Object {
      "cigar": "10M",
      "source": "21",
      "strand1": "+",
      "strand2": "+",
      "tags": Object {},
      "target": "23",
    },
    Object {
      "cigar": "10M",
      "source": "22",
      "strand1": "-",
      "strand2": "2",
      "tags": Object {},
      "target": "6",
    },
    Object {
      "cigar": "10M",
      "source": "24",
      "strand1": "+",
      "strand2": "4",
      "tags": Object {},
      "target": "7",
    },
    Object {
      "cigar": "10M",
      "source": "24",
      "strand1": "-",
      "strand2": "4",
      "tags": Object {},
      "target": "7",
    },
    Object {
      "cigar": "10M",
      "source": "3",
      "strand1": "+",
      "strand2": "+",
      "tags": Object {},
      "target": "4",
    },
    Object {
      "cigar": "10M",
      "source": "3",
      "strand1": "-",
      "strand2": "-",
      "tags": Object {},
      "target": "6",
    },
    Object {
      "cigar": "10M",
      "source": "3",
      "strand1": "-",
      "strand2": "-",
      "tags": Object {},
      "target": "8",
    },
    Object {
      "cigar": "10M",
      "source": "4",
      "strand1": "-",
      "strand2": "-",
      "tags": Object {},
      "target": "7",
    },
  ],
  "nodes": Array [
    Object {
      "id": "1",
      "length": 6871,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "10",
      "length": 251,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "11",
      "length": 208,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "12",
      "length": 186,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "16",
      "length": 157,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "18",
      "length": 145,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "19",
      "length": 134,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "2",
      "length": 4589,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "20",
      "length": 134,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "21",
      "length": 133,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "22",
      "length": 132,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "23",
      "length": 132,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "24",
      "length": 107,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "3",
      "length": 2044,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "4",
      "length": 1744,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "5",
      "length": 1378,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "6",
      "length": 1356,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "7",
      "length": 920,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "8",
      "length": 876,
      "sequence": "*",
      "tags": Object {},
    },
    Object {
      "id": "9",
      "length": 255,
      "sequence": "*",
      "tags": Object {},
    },
  ],
  "paths": Array [],
}
`;
